### DISCUSSION TITLE
Set global 'formatoptions' that doesn't get overwritten by runtime files

### DISCUSSION THREAD

**Discussion by Danielkonge** [neovim/neovim#26885] [Q&A] [answered]
In my init.lua (or really a file sourced by init.lua) I have:
```lua
vim.opt.formatoptions = "cqrnj" -- default: "tcqj"
```
But for most files I open (e.g. Lua files) `:lua= vim.opt.formatoptions:get()` gives
```lua
{
  c = true,
  j = true,
  l = true,
  n = true,
  o = true,
  q = true,
  r = true
}
```
and `:set verbose=1 formatoptions?` gives
```
formatoptions=njcroql
	Last set from /opt/homebrew/Cellar/neovim/HEAD-f69658b/share/nvim/runtime/ftplugin/lua.vim line 20
```
which makes sense since I can see `formatoptions` being set here: https://github.com/neovim/neovim/blob/444f37fe510f4c28c59bade40d7ba152a5ee8f7c/runtime/ftplugin/lua.vim#L20

But I would like my formatoptions to be global and not overwritten by the runtime files. What is the recommended way to do this?

So far I have been making sure my settings work as expected by using the following autocmd:
```lua
local daniel_format = vim.api.nvim_create_augroup('DanielFormat', { clear = true })
vim.api.nvim_create_autocmd("BufEnter", {
	group = daniel_format,
	pattern = "*",
	callback = function()
		vim.opt.formatoptions = "cqrnj"
	end
})
```

**Comment by wookayin** [comment]
```
setlocal formatoptions-=t formatoptions+=croql
```
These will change local `&formatoptions` only, not the global option.  You can probably see that `vim.opt_global.formatopions` would have remained unchanged. Also try `:setglobal formatoptions?`.

**Reply by Danielkonge** [reply]
I know they only change the local option, but the local option is what dictates the setting when I open e.g. a Lua file, so I am looking for a way to always use the global setting if possible. (My autocmd works fine for this, but I thought there might be an easier/better way.)

**Reply by Danielkonge** [reply]
Also, `:setglobal formatoptions?` gives
```lua
 formatoptions=cqrnj
```
as expected. This is not a bug, more of a question about the best way to do what I want.

**Reply by clason** [reply, marked as answer]
There might be more targeted events (`FileType`, `OptionSet`), but an autocommand -- or local `ftplugin` is The Way.

**Reply by Danielkonge** [reply]
I have tried `OptionSet` with pattern `formatoptions`, but couldn't get it to work (didn't try that hard to be fair). 

I will probably just keep my current autocmd from above or use `FileType` with pattern "*" (I guess that is more targeted correctly, since it is the runtime `ftplugin` files that change the setting), since these work fine. I just thought there might be some way to "lock" a setting globally. 

Thanks for the quick responses.

**Reply by clason** [reply]
No, you can't "lock" global settings.

**Reply by Danielkonge** [reply]
I don't know that much about the option code in general, but would it make sense to have something like `vim.opt_always.<option>` (maybe with a better name) that takes precedence over all current versions of `vim.opt.<option>` (`opt_local`, `opt_global` and `vim.wo`, `vim.bo`, ...)? 

This is not that important to me, so I don't really want to create a GitHub issue if it would be very hard to implement, or you think it doesn't really make sense to have in neovim.

**Reply by clason** [reply]
You don't know what you're asking....

Yes, this would be _very_ hard to implement, and likely to break all manner of Vim compatibility (important for plugins).

(The Lua code is just a more or less thin wrapper over the internal option handling, where this would have to be introduced.)

**Reply by Danielkonge** [reply]
> You don't know what you're asking....

Haha, yeah. I have looked at small parts of neovim's code when I looked at stuff I wanted to change, but I really have no wider perspective for how everything works together.

> Yes, this would be _very_ hard to implement, and likely to break all manner of Vim compatibility (important for plugins).
> 
> (The Lua code is just a more or less thin wrapper over the internal option handling, where this would have to be introduced.)

I will just not make a GitHub issue then. It doesn't sound like it would be worth the effort, since the current workaround is only a minor inconvenience.
